### PR TITLE
Annotating global declarations and entire specifications

### PR DESCRIPTION
**Pull request by Roman-Manevich**
* Two new chapters for annotating global declarations and entire ASL specifications.
* Introduced shorthand notation for environment updates.
* Fixed definition of domains, which in particularly handles under-constrained integers.

Test plan:
$ make test
$ dune runtest asllib

### CODE REVIEW

File: asllib/doc/ASLTypingReference.tex [hunk 1]
```diff
@@ -839,37 +839,40 @@ \section{TypingRule.Domain}
 
 \begin{emptyformal}
       \subsection{Formally}
+\newcommand\runtimeenv[0]{E}
```
[the remark this answers is not in the document]
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:844]: Right. There are two reasons for this:
1. An under-constrained integer represents a constant whose value is not known until runtime. Therefore, the domain needs to get a specific value to yield the (singleton) domain of that integer.
2. Similarly, config variables need a value to yield a domain.

File: asllib/doc/ASLTypingReference.tex [hunk 2]
```diff
@@ -5568,7 +5575,295 @@ \subsection{Comments}
...
+Type checking the specification is considered successful if all declarations can be successfully annotated.
+
+\newcommand\decls[0]{\texttt{decls}}
+
+\section{TypingRule.Specification \label{sec:TypingRule.Specification}}
+\subsection{Prose}
+Annotating an ASL specification \texttt{decls} in an environment \texttt{env} results in a rewritten specification \texttt{decls'}
+and a new environment \texttt{new\_env} and all of the following apply:
+\begin{itemize}
+  \item \texttt{ordered\_decls} is the result of topologically ordering \texttt{decls} according to their
```
[the remark this answers is not in the document]
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:5840]: What is the difference specifically?

File: asllib/doc/ASLTypingReference.tex [hunk 3]
```diff
@@ -5568,7 +5575,295 @@ \subsection{Comments}
...
+\end{emptyformal}
+\subsection{Comments}
+
+This relates to \identi{HJRD}, \identd{BTBR}, \identi{FSFQ}, \identi{PFGQ}, and \identr{PGFC}.
+
+\section{TypingRule.DeclareGlobalStorage \label{sec:TypingRule.DeclareGlobalStorage}}
+\subsection{Prose}
+Annotating a global storage declaration \texttt{d} in a given environment \texttt{env} results
+in \texttt{new\_env} and one of the following applies:
+\begin{enumerate}[label=Case \arabic*:]
```
[the remark this answers is not in the document]
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:5666]: Sure. I did want you to see this suggestion before applying all over (in the next PR).

File: asllib/doc/ASLTypingReference.tex [hunk 4]
```diff
@@ -5568,7 +5575,295 @@ \subsection{Comments}
...
+    \item An error ``identifier already declared'' is raised.
+  \end{itemize}
+  \item All of the following apply:
+  \begin{itemize}
+    \item \name\ is not declared in \tenv.
+    \item  \tenv',\tty' are the result of attempting to add \texttt{name} as a subtype of \texttt{sup} and 
+    constructing the type \tty' as \tty\ with the added fields \texttt{fields} in \tenv.
+    \item \tty' is a valid type.
+    \item \tenv'' is \tenv' extended with \tty'.
+    \item If \tty' has the structure of an enumeration than \newenv\ is \tenv'' extended with the declarations of constants for each identifier,
```
[the remark this answers is not in the document]
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:5795]: Right, but I'm thinking of deferring this to the next PR. I don't want to risk not merging what's already in this PR.

File: asllib/doc/ASLTypingReference.tex [hunk 5]
```diff
@@ -5568,7 +5575,295 @@ \subsection{Comments}
...
+\newcommand\hassubprogramtypeclash[0]{\texttt{subprogram\_type\_clash}}
+\newcommand\subprogramclash[0]{\texttt{subprogram\_clash}}
+\newcommand\argsclash[0]{\texttt{args\_clash}}
+
+We define the helper rules \hassubprogramtypeclash, \argsclash, and subprogramclash\ to determined whether
+two subprogram clash in terms of their subprogram types and their lists of argument types,
+and then use them to annotate a subprogram declaration.
+
+\begin{mathpar}
+\inferrule{}{\hassubprogramtypeclash(\tenv, \texttt{ST\_Function}, \Ignore)}
```
[the remark this answers is not in the document]
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex:5619]: RIght. In the next PR.

File: asllib/doc/ASLTypingReference.tex (hunk 1 above)
[the remark this answers is not in the document]
Comment (Roman-Manevich) [reply asllib/doc/ASLTypingReference.tex]: 1. Sure.
2. Yes, but the value of the config may determine the length of a bitvector/array, right? (Not completely sure, but I think so.)

**Comment by jalglave** [comment]
Thank you both.
Jade

**Pull request merged**